File: android/src/main/java/io/hyperswitch/capacitor/HyperswitchPlugin.java
package io.hyperswitch.capacitor;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Map;

@CapacitorPlugin(name = "Hyperswitch")
public class HyperswitchPlugin extends Plugin {

    private final HyperswitchImpl implementation = HyperswitchImpl.getInstance();

    /**
     * Called once after the plugin is registered with Capacitor.
     * Registers a NativeEventListener so that widget events (FORM_STATUS, CVC_STATUS, etc.)
     * are forwarded to the JS layer via notifyListeners("paymentEvent", ...).
     */
    @Override
    public void load() {
        implementation.setEventListener((type, payload, source) -> {
            JSObject data = new JSObject();
            data.put("type", type);
            JSObject payloadJs = new JSObject();
            for (Map.Entry<String, Object> entry : payload.entrySet()) {
                Object value = entry.getValue();
                if (value != null) {
                    payloadJs.put(entry.getKey(), value.toString());
                }
            }
            data.put("payload", payloadJs);
            String channel = switch (source) {
                case "paymentElement" -> "paymentElementEvent";
                case "cvcWidget" -> "cvcWidgetEvent";
                default -> "paymentEvent";
            };
            notifyListeners(channel, data);
        });
    }

    @PluginMethod
    public void init(PluginCall call) {
        String publishableKey = call.getString("publishableKey");

        if (publishableKey == null || publishableKey.isEmpty()) {
            call.reject("publishableKey is required");
            return;
        }

        String profileId = call.getString("profileId");
        JSObject customConfig = call.getObject("customConfig");
        String environment = call.getString("environment", "PROD");

        implementation.init(getActivity(), publishableKey, profileId, customConfig, environment);
        call.resolve();
    }

    @PluginMethod
    public void elements(PluginCall call) {
        call.setKeepAlive(true);
        JSObject elementsOptions = call.getObject("elementsOptions");
        implementation.elements(elementsOptions, new HyperswitchImpl.ElementsCallback() {
            @Override
            public void onReady(String handlerId) {
                JSObject result = new JSObject();
                result.put("handlerId", handlerId);
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void createElement(PluginCall call) {
        String type = call.getString("type");
        JSObject createOptions = call.getObject("createOptions");
        implementation.createElement(type, createOptions);
        call.resolve();
    }

    @PluginMethod
    public void updateIntent(PluginCall call) {
        call.setKeepAlive(true);
        String sdkAuthorization = call.getString("sdkAuthorization");
        implementation.updateIntent(sdkAuthorization, new HyperswitchImpl.PaymentResultCallback() {
            @Override
            public void onResult(JSObject result) {
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void initPaymentSession(PluginCall call) {
        call.setKeepAlive(true);
        JSObject paymentSessionOptions = call.getObject("paymentSessionOptions");
        implementation.initPaymentSession(paymentSessionOptions, new HyperswitchImpl.InitPaymentSessionCallback() {
            @Override
            public void onReady() {
                call.setKeepAlive(false);
                call.resolve();
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void getCustomerSavedPaymentMethods(PluginCall call) {
        call.setKeepAlive(true);
        implementation.getCustomerSavedPaymentMethods(new HyperswitchImpl.CustomerSavedPaymentMethodsCallback() {
            @Override
            public void onReady(String handlerId) {
                JSObject result = new JSObject();
                result.put("handlerId", handlerId);
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void getCustomerSavedPaymentMethodData(PluginCall call) {
        String handlerId = call.getString("handlerId");
        if (handlerId == null || handlerId.isEmpty()) { call.reject("handlerId is required"); return; }
        call.resolve(implementation.getCustomerSavedPaymentMethodData(handlerId));
    }

    @PluginMethod
    public void getCustomerDefaultSavedPaymentMethodData(PluginCall call) {
        String handlerId = call.getString("handlerId");
        if (handlerId == null || handlerId.isEmpty()) { call.reject("handlerId is required"); return; }
        call.resolve(implementation.getCustomerDefaultSavedPaymentMethodData(handlerId));
    }

    @PluginMethod
    public void getCustomerLastUsedPaymentMethodData(PluginCall call) {
        String handlerId = call.getString("handlerId");
        if (handlerId == null || handlerId.isEmpty()) { call.reject("handlerId is required"); return; }
        call.resolve(implementation.getCustomerLastUsedPaymentMethodData(handlerId));
    }

    @PluginMethod
    public void presentPaymentSheet(PluginCall call) {
        call.setKeepAlive(true);
        JSObject sheetOptions = call.getObject("sheetOptions");
        implementation.presentPaymentSheet(sheetOptions, new HyperswitchImpl.PaymentSheetCallback() {
            @Override
            public void onResult(JSObject result) {
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(Exception e) {
                call.setKeepAlive(false);
                call.reject(e.getMessage(), e);
            }
        });
    }

    @PluginMethod
    public void confirmPayment(PluginCall call) {
        call.setKeepAlive(true);
        JSObject confirmParams = call.getObject("confirmParams");
        implementation.confirmPayment(confirmParams, new HyperswitchImpl.PaymentResultCallback() {
            @Override
            public void onResult(JSObject result) {
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void confirmWithCustomerDefaultPaymentMethod(PluginCall call) {
        String handlerId = call.getString("handlerId");
        if (handlerId == null || handlerId.isEmpty()) { call.reject("handlerId is required"); return; }
        call.setKeepAlive(true);
        implementation.confirmWithCustomerDefaultPaymentMethod(handlerId, new HyperswitchImpl.PaymentResultCallback() {
            @Override
            public void onResult(JSObject result) {
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    @PluginMethod
    public void confirmWithCustomerLastUsedPaymentMethod(PluginCall call) {
        String handlerId = call.getString("handlerId");
        if (handlerId == null || handlerId.isEmpty()) { call.reject("handlerId is required"); return; }
        call.setKeepAlive(true);
        implementation.confirmWithCustomerLastUsedPaymentMethod(handlerId, new HyperswitchImpl.PaymentResultCallback() {
            @Override
            public void onResult(JSObject result) {
                call.setKeepAlive(false);
                call.resolve(result);
            }

            @Override
            public void onError(String message) {
                call.setKeepAlive(false);
                call.reject(message);
            }
        });
    }

    // ── PaymentElement lifecycle stubs ────────────────────────────────────────

    @PluginMethod
    public void elementMount(PluginCall call) {
        // Mounting is handled natively via PaymentElementPlugin.create() + createElement().
        call.resolve();
    }

    @PluginMethod
    public void elementOn(PluginCall call) {
        // Event subscriptions are registered natively when the element is bound (createElement).
        // Events are pushed to JS via notifyListeners("paymentEvent", ...).
        // This method is kept as a no-op so the JS bridge doesn't reject the call.
        call.resolve();
    }

    @PluginMethod
    public void elementCollapse(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementBlur(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementFocus(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementClear(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementUpdate(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementUnmount(PluginCall call) {
        call.resolve();
    }

    @PluginMethod
    public void elementDestroy(PluginCall call) {
        call.resolve();
    }
}

File: ios/Sources/HyperswitchPlugin/HyperswitchPlugin.swift
import Foundation
import Capacitor

@objc(HyperswitchPlugin)
public class HyperswitchPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "HyperswitchPlugin"
    public let jsName = "Hyperswitch"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "init",                                    returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elements",                                returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createElement",                           returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "updateIntent",                            returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "initPaymentSession",                      returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "presentPaymentSheet",                     returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCustomerSavedPaymentMethods",          returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCustomerSavedPaymentMethodData",       returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCustomerDefaultSavedPaymentMethodData",returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getCustomerLastUsedPaymentMethodData",    returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "confirmWithCustomerDefaultPaymentMethod", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "confirmWithCustomerLastUsedPaymentMethod",returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "confirmPayment",                          returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementOn",                               returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementCollapse",                         returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementBlur",                             returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementUpdate",                           returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementDestroy",                          returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementUnmount",                          returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementMount",                            returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementFocus",                            returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "elementClear",                            returnType: CAPPluginReturnPromise),
    ]

    private let impl = HyperswitchImpl.shared

    // ── Init ───────────────────────────────────────────────────────────────────

    @objc func `init`(_ call: CAPPluginCall) {
        guard let publishableKey = call.getString("publishableKey"), !publishableKey.isEmpty else {
            call.reject("publishableKey is required")
            return
        }

        guard let viewController = bridge?.viewController else {
            call.reject("Unable to get view controller from bridge")
            return
        }

        let profileId    = call.getString("profileId")
        let customConfig = call.getObject("customConfig")
        let environment  = call.getString("environment") ?? "PROD"

        impl.initialize(
            viewController: viewController,
            publishableKey: publishableKey,
            profileId: profileId,
            customConfig: customConfig,
            environment: environment
        )
        call.resolve()
    }

    // ── Elements ───────────────────────────────────────────────────────────────

    @objc func elements(_ call: CAPPluginCall) {
        let options = call.getObject("elementsOptions")
        let sdkAuthorization = options?["sdkAuthorization"] as? String ?? ""

        impl.elements(
            sdkAuthorization: sdkAuthorization,
            onReady: { handlerId in call.resolve(["handlerId": handlerId]) },
            onError: { msg in call.reject(msg) }
        )
    }

    @objc func createElement(_ call: CAPPluginCall) {
        let type          = call.getString("type") ?? ""
        let createOptions = call.getObject("createOptions")
        impl.createElement(type: type, createOptions: createOptions)
        call.resolve()
    }

    @objc func updateIntent(_ call: CAPPluginCall) {
        let sdkAuthorization = call.getString("sdkAuthorization") ?? ""

        impl.updateIntent(
            sdkAuthorization: sdkAuthorization,
            onResult: { result in call.resolve(result) },
            onError:  { msg in call.reject(msg) }
        )
    }

    // ── InitPaymentSession ─────────────────────────────────────────────────────

    @objc func initPaymentSession(_ call: CAPPluginCall) {
        let options          = call.getObject("paymentSessionOptions")
        let sdkAuthorization = options?["sdkAuthorization"] as? String ?? ""

        impl.initPaymentSession(
            sdkAuthorization: sdkAuthorization,
            onReady: { call.resolve() },
            onError: { msg in call.reject(msg) }
        )
    }

    @objc func presentPaymentSheet(_ call: CAPPluginCall) {
        guard let viewController = bridge?.viewController else {
            call.reject("Unable to get view controller from bridge")
            return
        }

        impl.presentPaymentSheet(
            viewController: viewController,
            onResult: { result in call.resolve(result) },
            onError:  { msg in call.reject(msg) }
        )
    }

    // ── Saved payment methods ──────────────────────────────────────────────────

    @objc func getCustomerSavedPaymentMethods(_ call: CAPPluginCall) {
        impl.getCustomerSavedPaymentMethods(
            onReady: { handlerId in call.resolve(["handlerId": handlerId]) },
            onError: { msg in call.reject(msg) }
        )
    }

    @objc func getCustomerSavedPaymentMethodData(_ call: CAPPluginCall) {
        guard let handlerId = call.getString("handlerId"), !handlerId.isEmpty else {
            call.reject("handlerId is required"); return
        }
        call.resolve(impl.getCustomerSavedPaymentMethodData(handlerId: handlerId))
    }

    @objc func getCustomerDefaultSavedPaymentMethodData(_ call: CAPPluginCall) {
        guard let handlerId = call.getString("handlerId"), !handlerId.isEmpty else {
            call.reject("handlerId is required"); return
        }
        call.resolve(impl.getCustomerDefaultSavedPaymentMethodData(handlerId: handlerId))
    }

    @objc func getCustomerLastUsedPaymentMethodData(_ call: CAPPluginCall) {
        guard let handlerId = call.getString("handlerId"), !handlerId.isEmpty else {
            call.reject("handlerId is required"); return
        }
        call.resolve(impl.getCustomerLastUsedPaymentMethodData(handlerId: handlerId))
    }

    // ── Confirm with saved methods ─────────────────────────────────────────────

    @objc func confirmWithCustomerDefaultPaymentMethod(_ call: CAPPluginCall) {
        guard let handlerId = call.getString("handlerId"), !handlerId.isEmpty else {
            call.reject("handlerId is required"); return
        }
        impl.confirmWithCustomerDefaultPaymentMethod(
            handlerId: handlerId,
            onResult: { result in call.resolve(result) },
            onError:  { msg in call.reject(msg) }
        )
    }

    @objc func confirmWithCustomerLastUsedPaymentMethod(_ call: CAPPluginCall) {
        guard let handlerId = call.getString("handlerId"), !handlerId.isEmpty else {
            call.reject("handlerId is required"); return
        }
        impl.confirmWithCustomerLastUsedPaymentMethod(
            handlerId: handlerId,
            onResult: { result in call.resolve(result) },
            onError:  { msg in call.reject(msg) }
        )
    }

    // ── confirmPayment ─────────────────────────────────────────────────────────

    @objc func confirmPayment(_ call: CAPPluginCall) {
        let confirmParams = call.getObject("confirmParams")

        impl.confirmPayment(
            confirmParams: confirmParams,
            onResult: { result in call.resolve(result) },
            onError:  { msg in call.reject(msg) }
        )
    }

    // ── PaymentElement lifecycle stubs ─────────────────────────────────────────
    // These are no-ops on iOS; mounting/unmounting is handled natively by
    // PaymentElementPlugin. They resolve immediately so JS callers don't hang.

    @objc func elementMount(_ call: CAPPluginCall)    { call.resolve() }
    @objc func elementOn(_ call: CAPPluginCall)       { call.resolve() }
    @objc func elementCollapse(_ call: CAPPluginCall) { call.resolve() }
    @objc func elementBlur(_ call: CAPPluginCall)     { call.resolve() }
    @objc func elementFocus(_ call: CAPPluginCall)    { call.resolve() }
    @objc func elementClear(_ call: CAPPluginCall)    { call.resolve() }
    @objc func elementUpdate(_ call: CAPPluginCall)   { call.resolve() }
    @objc func elementUnmount(_ call: CAPPluginCall)  { call.resolve() }
    @objc func elementDestroy(_ call: CAPPluginCall)  { call.resolve() }
}
